// arjs-studio-backend/dist/providers/github.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GithubProvider = void 0;
var rest_1 = require("@octokit/rest");
var provider_1 = require("./provider");
var index_1 = require("../types/index");
/* eslint-disable no-unused-vars */
var MODE_FILE = '100644';
var MODE_EXEC = '100755';
var MODE_SUBDIR = '040000';
var MODE_SUBMOD = '160000';
var MODE_SYMLINK = '120000';
var TYPE_BLOB = 'blob';
var TYPE_TREE = 'tree';
var TYPE_COMMIT = 'commit';
var DEFAULT_PAGE_BRANCH = 'gh-pages';
var DEFAULT_COMMIT_MESSAGE = 'built with AR.js Studio';
var rndRepoName = function () {
    var max = 999999;
    var min = 100000;
    return "arjs-studio-".concat(Math.floor(Math.random() * (max - min)) + min);
};
var GithubProvider = /** @class */ (function (_super) {
    __extends(GithubProvider, _super);
    function GithubProvider() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * Add file to bundle.
     *
     * @param {string} path repository path
     * @param {string} content
     * @param {string} [encoding] utf-8 or base64
     */
    GithubProvider.prototype.addFile = function (path, content, encoding) {
        if (encoding === void 0) { encoding = index_1.EncodingType.ENC_UTF8; }
        this.files.push({
            content: content,
            encoding: encoding,
            path: path,
        });
    };
    /**
     * Publish files creating a commit in a custom branch.
     *
     * @param {Object} config
     * @param {string} config.token - GitHub authorization token
     * @param {string} [config.message] - commit message
     * @param {string|null} [config.repo] - repository name (default to random name)
     * @param {string|null} [config.branch] - branch name (default to 'gh-pages')
     * @param {string|null} [config.owner] - owner login name (automatically retrieved if null)
     * @return {Promise<string|null>} - Pages URL for this repository
     */
    GithubProvider.prototype.serveFiles = function (_a) {
        var token = _a.token, _b = _a.message, message = _b === void 0 ? DEFAULT_COMMIT_MESSAGE : _b, _c = _a.repo, repo = _c === void 0 ? null : _c, _d = _a.branch, branch = _d === void 0 ? null : _d, _e = _a.owner, owner = _e === void 0 ? null : _e;
        return __awaiter(this, void 0, void 0, function () {
            var _f, _g, ghBranch, blobFiles, tree, commit;
            var _this = this;
            return __generator(this, function (_h) {
                switch (_h.label) {
                    case 0: return [4 /*yield*/, _super.prototype.serveFiles.call(this)];
                    case 1:
                        _h.sent();
                        if (!token) {
                            throw new Error('Missing required token parameter');
                        }
                        this.client = new rest_1.Octokit({
                            auth: token,
                        });
                        _f = this;
                        _g = owner;
                        if (_g) return [3 /*break*/, 3];
                        return [4 /*yield*/, this.getOwner()];
                    case 2:
                        _g = (_h.sent());
                        _h.label = 3;
                    case 3:
                        _f.owner = _g;
                        this.repo = repo || rndRepoName();
                        this.branch = branch || index_1.Branch.PAGES;
                        return [4 /*yield*/, this.getOrCreateRepo(this.repo)];
                    case 4:
                        _h.sent();
                        return [4 /*yield*/, this.getOrCreateBranch(this.branch)];
                    case 5:
                        ghBranch = _h.sent();
                        return [4 /*yield*/, Promise.all(
                            /** @type {Array<GitFile>} */
                            this.files.map(function (file) {
                                return _this.createBlob(file.content, file.encoding).then(function (data) { return ({
                                    path: file.path,
                                    sha: data.sha,
                                    mode: MODE_FILE,
                                    type: TYPE_BLOB,
                                }); });
                            }))];
                    case 6:
                        blobFiles = _h.sent();
                        return [4 /*yield*/, this.createTree(blobFiles)];
                    case 7:
                        tree = _h.sent();
                        return [4 /*yield*/, this.createCommit(message, tree.sha, [
                                ghBranch.commit.sha,
                            ])];
                    case 8:
                        commit = _h.sent();
                        return [4 /*yield*/, this.updateRef(commit.sha, this.branch)];
                    case 9:
                        _h.sent();
                        if (!(this.branch !== DEFAULT_PAGE_BRANCH)) return [3 /*break*/, 11];
                        return [4 /*yield*/, this.enablePages(this.branch)];
                    case 10:
                        _h.sent();
                        _h.label = 11;
                    case 11: return [4 /*yield*/, this.rebuildPages()];
                    case 12:
                        _h.sent();
                        return [2 /*return*/, this.getPagesUrl()];
                }
            });
        });
    };
    /**
     * @return {Promise<string>}
     */
    GithubProvider.prototype.getOwner = function () {
        return __awaiter(this, void 0, void 0, function () {
            var data;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.client.users.getAuthenticated()];
                    case 1:
                        data = (_a.sent()).data;
                        return [2 /*return*/, data.login];
                }
            });
        });
    };
    GithubProvider.prototype.getOrCreateRepo = function (name) {
        return __awaiter(this, void 0, void 0, function () {
            var e_1;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 4]);
                        return [4 /*yield*/, this.getRepo(name)];
                    case 1: return [2 /*return*/, _a.sent()];
                    case 2:
                        e_1 = _a.sent();
                        return [4 /*yield*/, this.createRepo(name)];
                    case 3: 
                    // console.warn('repo not found, creating');
                    return [2 /*return*/, _a.sent()];
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    GithubProvider.prototype.getOrCreateBranch = function (name) {
        return __awaiter(this, void 0, void 0, function () {
            var e_2, masterRef;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 6]);
                        return [4 /*yield*/, this.getBranch(name)];
                    case 1: return [2 /*return*/, _a.sent()];
                    case 2:
                        e_2 = _a.sent();
                        return [4 /*yield*/, this.getRef(index_1.Branch.MASTER)];
                    case 3:
                        masterRef = _a.sent();
                        return [4 /*yield*/, this.createBranch(masterRef.object.sha, name)];
                    case 4:
                        _a.sent();
                        return [4 /*yield*/, this.getBranch(name)];
                    case 5: return [2 /*return*/, _a.sent()];
                    case 6: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Scope: public_repo
     *
     * @link https://developer.github.com/v3/repos/#create
     *
     * @param {string} name repository name
     * @return {Promise<Octokit.ReposCreateForAuthenticatedUserResponse>}
     */
    GithubProvider.prototype.createRepo = function (name) {
        return this.client.repos
            .createForAuthenticatedUser({
            name: name,
            auto_init: true, // first commit
        })
            .then(function (_a) {
            var data = _a.data;
            return data;
        });
    };
    /**
     * @param {string} name repository name
     * @return {Promise<Octokit.ReposGetResponse>}
     */
    GithubProvider.prototype.getRepo = function (name) {
        return __awaiter(this, void 0, void 0, function () {
            var data;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.client.repos.get({
                            owner: this.owner,
                            repo: name,
                        })];
                    case 1:
                        data = (_a.sent()).data;
                        return [2 /*return*/, data];
                }
            });
        });
    };
    /**
     * @link https://developer.github.com/v3/git/refs/#create-a-reference
     *
     * @param {string} sha commit SHA from which to branch off
     * @param {string} [branch] new branch name (default to Pages branch)
     * @return {Promise<Octokit.GitCreateRefResponse>}
     */
    GithubProvider.prototype.createBranch = function (sha, branch) {
        if (branch === void 0) { branch = DEFAULT_PAGE_BRANCH; }
        return __awaiter(this, void 0, void 0, function () {
            var data;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.client.git.createRef({
                            owner: this.owner,
                            repo: this.repo,
                            ref: "refs/heads/".concat(branch),
                            sha: sha,
                        })];
                    case 1:
                        data = (_a.sent()).data;
                        return [2 /*return*/, data];
                }
            });
        });
    };
    /**
     * @link https://developer.github.com/v3/repos/branches/#get-branch
     *
     * @param name
     * @return {Promise<Octokit.ReposGetBranchResponse>}
     */
    GithubProvider.prototype.getBranch = function (name) {
        return __awaiter(this, void 0, void 0, function () {
            var data;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.client.repos.getBranch({
                            owner: this.owner,
                            repo: this.repo,
                            branch: name,
                        })];
                    case 1:
                        data = (_a.sent()).data;
                        return [2 /*return*/, data];
                }
            });
        });
    };
    /**
     * Get SHA of currently referenced commit for a branch.
     * Useful for getting a reference to create a new branch.
     *
     * @link https://developer.github.com/v3/git/refs/#get-a-single-reference
     *
     * @param {string} [branch] branch name (default to master)
     * @return {Promise<Octokit.GitGetRefResponse>}
     */
    GithubProvider.prototype.getRef = function (branch) {
        if (branch === void 0) { branch = index_1.Branch.MASTER; }
        return __awaiter(this, void 0, void 0, function () {
            var data;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.client.git.getRef({
                            owner: this.owner,
                            repo: this.repo,
                            ref: "heads/".concat(branch),
                        })];
                    case 1:
                        data = (_a.sent()).data;
                        return [2 /*return*/, data];
                }
            });
        });
    };
    /**
     * @param {string} [branch] branch on which to enable Pages (default to Pages branch)
     * @return {Promise<Octokit.Response<Octokit.ReposEnablePagesSiteResponse>>}
     */
    GithubProvider.prototype.enablePages = function (branch) {
        if (branch === void 0) { branch = this.branch; }
        return this.client.repos.enablePagesSite({
            owner: this.owner,
            repo: this.repo,
            source: {
                branch: branch,
            },
        });
    };
    /**
     * Force a Pages build.
     *
     * @return {Promise<Octokit.ReposRequestPageBuildResponse>}
     */
    GithubProvider.prototype.rebuildPages = function () {
        return __awaiter(this, void 0, void 0, function () {
            var data;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.client.repos.requestPageBuild({
                            owner: this.owner,
                            repo: this.repo,
                        })];
                    case 1:
                        data = (_a.sent()).data;
                        return [2 /*return*/, data];
                }
            });
        });
    };
    /**
     * @return {Promise<string>}
     */
    GithubProvider.prototype.getPagesUrl = function () {
        return __awaiter(this, void 0, void 0, function () {
            var data;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.client.repos.getPages({
                            owner: this.owner,
                            repo: this.repo,
                        })];
                    case 1:
                        data = (_a.sent()).data;
                        return [2 /*return*/, data.html_url];
                }
            });
        });
    };
    /**
     * Add a single file with a commit.
     *
     * @param {string} content file content
     * @param {string} path file path
     * @param {string} message commit message
     * @param {string} [branch] branch to commit to (default to Pages branch)
     * @return {Promise<Octokit.ReposCreateOrUpdateFileResponse>}
     */
    GithubProvider.prototype.commitFile = function (content, path, message, branch) {
        if (branch === void 0) { branch = this.branch; }
        return __awaiter(this, void 0, void 0, function () {
            var data;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.client.repos.createOrUpdateFile({
                            owner: this.owner,
                            repo: this.repo,
                            path: path,
                            message: message,
                            content: content,
                            branch: branch,
                        })];
                    case 1:
                        data = (_a.sent()).data;
                        return [2 /*return*/, data];
                }
            });
        });
    };
    /**
     * Create a file tree to commit.
     *
     * @link https://octokit.github.io/rest.js/#octokit-routes-git-create-tree
     * @link https://developer.github.com/v3/git/trees/#create-a-tree
     *
     * @param {GitFile[]} files list of blobs
     * @return {Promise<Octokit.GitCreateTreeResponse>}
     */
    GithubProvider.prototype.createTree = function (files) {
        return __awaiter(this, void 0, void 0, function () {
            var data;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.client.git.createTree({
                            owner: this.owner,
                            repo: this.repo,
                            tree: files,
                        })];
                    case 1:
                        data = (_a.sent()).data;
                        return [2 /*return*/, data];
                }
            });
        });
    };
    /**
     * Create a file blob for a tree.
     *
     * @link https://developer.github.com/v3/git/blobs/#create-a-blob
     *
     * @param {string} content text or base64 blob content
     * @param [encoding] blob encoding, either 'utf-8' or 'base64' (default to 'utf-8')
     * @return {Promise<Octokit.GitCreateBlobResponse>}
     */
    GithubProvider.prototype.createBlob = function (content, encoding) {
        if (encoding === void 0) { encoding = provider_1.ENC_UTF8; }
        return __awaiter(this, void 0, void 0, function () {
            var data;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.client.git.createBlob({
                            owner: this.owner,
                            repo: this.repo,
                            content: content,
                            encoding: encoding,
                        })];
                    case 1:
                        data = (_a.sent()).data;
                        return [2 /*return*/, data];
                }
            });
        });
    };
    /**
     * Commit a tree.
     *
     * @link https://octokit.github.io/rest.js/#octokit-routes-git-create-commit
     * @link https://developer.github.com/v3/git/commits/#create-a-commit
     *
     * @param {string} message commit message
     * @param {string} tree SHA of the tree to commit
     * @param {string[]} [parents] array of parent commit SHA strings
     * @return {Promise<Octokit.GitCreateCommitResponse>}
     */
    GithubProvider.prototype.createCommit = function (message, tree, parents) {
        if (parents === void 0) { parents = []; }
        return __awaiter(this, void 0, void 0, function () {
            var data;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.client.git.createCommit({
                            owner: this.owner,
                            repo: this.repo,
                            message: message,
                            tree: tree,
                            parents: parents,
                        })];
                    case 1:
                        data = (_a.sent()).data;
                        return [2 /*return*/, data];
                }
            });
        });
    };
    /**
     * Update a branch to reference a new commit.
     *
     * @link https://octokit.github.io/rest.js/#octokit-routes-git-update-ref
     * @link https://developer.github.com/v3/git/refs/#update-a-reference
     *
     * @param {string} sha git commit SHA to reference
     * @param {string} [branch] branch to update (default to Pages branch)
     * @return {Promise<Octokit.GitUpdateRefResponse>}
     */
    GithubProvider.prototype.updateRef = function (sha, branch) {
        if (branch === void 0) { branch = this.branch; }
        return __awaiter(this, void 0, void 0, function () {
            var data;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.client.git.updateRef({
                            owner: this.owner,
                            repo: this.repo,
                            ref: "heads/".concat(branch),
                            sha: sha,
                            force: true, // ignore errors
                        })];
                    case 1:
                        data = (_a.sent()).data;
                        return [2 /*return*/, data];
                }
            });
        });
    };
    return GithubProvider;
}(provider_1.Provider));
exports.GithubProvider = GithubProvider;
