// arjs-studio-backend/dist/providers/zip.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ZipProvider = exports.ZIP_TYPE_NODE_BUFFER = exports.ZIP_TYPE_BLOB = exports.ZIP_TYPE_ARRAY_BUFFER = exports.ZIP_TYPE_UINT8_ARRAY = exports.ZIP_TYPE_ARRAY = exports.ZIP_TYPE_BINARY_STRING = exports.ZIP_TYPE_BASE64 = exports.ENC_BINARY = void 0;
var index_1 = require("../types/index");
var provider_1 = require("./provider");
var jszip_1 = __importDefault(require("jszip"));
exports.ENC_BINARY = 'binary';
exports.ZIP_TYPE_BASE64 = 'base64';
exports.ZIP_TYPE_BINARY_STRING = 'binarystring';
exports.ZIP_TYPE_ARRAY = 'array'; // array of bytes (0-255)
exports.ZIP_TYPE_UINT8_ARRAY = 'uint8array';
exports.ZIP_TYPE_ARRAY_BUFFER = 'arraybuffer';
exports.ZIP_TYPE_BLOB = 'blob';
exports.ZIP_TYPE_NODE_BUFFER = 'nodebuffer';
var COMPRESS_DEFLATE = 'DEFLATE';
var COMPRESS_STORE = 'STORE';
var ZipProvider = /** @class */ (function (_super) {
    __extends(ZipProvider, _super);
    function ZipProvider() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * Add file to bundle.
     *
     * @param {string} path file path inside the zip
     * @param {string|ArrayBuffer|Uint8Array|Buffer|Blob|Promise|Stream} content
     * @param {string} [encoding] utf-8 or base64
     */
    ZipProvider.prototype.addFile = function (path, content, encoding) {
        if (encoding === void 0) { encoding = index_1.EncodingType.ENC_UTF8; }
        this.files.push({
            content: content,
            encoding: encoding,
            path: path,
        });
    };
    /**
     * Creates a ZIP with all files.
     *
     * @param {string} [type] - output ZIP file format, default to base64 string
     * @param {number} [compress] - compression level, default to 0
     * @return {Promise<string|Array<number>|Uint8Array|ArrayBuffer|Blob|Buffer>} - ZIP file in chosen format
     */
    ZipProvider.prototype.serveFiles = function (_a) {
        var _b = _a.type, type = _b === void 0 ? index_1.EncodingType.ENC_UTF8 : _b, _c = _a.compress, compress = _c === void 0 ? 0 : _c;
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_d) {
                switch (_d.label) {
                    case 0: return [4 /*yield*/, _super.prototype.serveFiles.call(this)];
                    case 1:
                        _d.sent();
                        this.zip = new jszip_1.default();
                        if ([
                            exports.ZIP_TYPE_ARRAY_BUFFER,
                            exports.ZIP_TYPE_UINT8_ARRAY,
                            exports.ZIP_TYPE_BLOB,
                            exports.ZIP_TYPE_NODE_BUFFER,
                        ].indexOf(type) !== -1) {
                            if (!jszip_1.default.support[type]) {
                                throw new Error("Output type not supported by browser: ".concat(type));
                            }
                        }
                        this.files.forEach(function (file) {
                            _this.zip.file(file.path, file.content, __assign({ base64: file.encoding === provider_1.ENC_BASE64 }, (file.encoding === exports.ENC_BINARY && { binary: true })));
                        });
                        return [2 /*return*/, this.zip.generateAsync({
                                type: type,
                                compression: compress > 0 ? COMPRESS_DEFLATE : COMPRESS_STORE,
                                compressionOptions: {
                                    level: compress, // ignored if STORE
                                },
                            })];
                }
            });
        });
    };
    return ZipProvider;
}(provider_1.Provider));
exports.ZipProvider = ZipProvider;
