// arjs-studio-backend/dist/modules/marker/tools/barcode-marker-generator.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BarcodeMarkerGenerator = exports.MATRIX_5X5_BCH_22125 = exports.MATRIX_5X5_BCH_2277 = exports.MATRIX_4X4_BCH_1393 = exports.MATRIX_4X4_BCH_1355 = exports.MATRIX_3X3_PARITY_65 = exports.MATRIX_3X3_HAMMING_63 = void 0;
/// @ts-nocheck
var encoders = __importStar(require("./matrix-encoders"));
exports.MATRIX_3X3_HAMMING_63 = '3x3_hamming_6_3';
exports.MATRIX_3X3_PARITY_65 = '3x3_parity_6_5';
exports.MATRIX_4X4_BCH_1355 = '4x4_bch_13_5_5';
exports.MATRIX_4X4_BCH_1393 = '4x4_bch_13_9_3';
exports.MATRIX_5X5_BCH_2277 = '5x5_bch_22_7_7';
exports.MATRIX_5X5_BCH_22125 = '5x5_bch_22_12_5';
var BARCODE_MATRIX_TYPES = [
    {
        id: exports.MATRIX_3X3_HAMMING_63,
        maxNumMarkers: 8,
        hamming: 3,
        matrixSize: 3,
        encoder: encoders.encode_3x3_hamming_6_3,
    },
    {
        id: exports.MATRIX_3X3_PARITY_65,
        maxNumMarkers: 32,
        hamming: 1,
        matrixSize: 3,
        encoder: encoders.encode_3x3_parity_6_5,
    },
    {
        id: exports.MATRIX_4X4_BCH_1355,
        maxNumMarkers: 32,
        hamming: 5,
        matrixSize: 4,
        encoder: encoders.encode_4x4_bch_13_5_5,
    },
    {
        id: exports.MATRIX_4X4_BCH_1393,
        maxNumMarkers: 512,
        hamming: 3,
        matrixSize: 4,
        encoder: encoders.encode_4x4_bch_13_9_3,
    },
    {
        id: exports.MATRIX_5X5_BCH_2277,
        maxNumMarkers: 128,
        hamming: 7,
        matrixSize: 5,
        encoder: encoders.encode_5x5_bch_22_7_7,
    },
    {
        id: exports.MATRIX_5X5_BCH_22125,
        maxNumMarkers: 4096,
        hamming: 5,
        matrixSize: 5,
        encoder: encoders.encode_5x5_bch_22_12_5,
    },
];
var BarcodeMarkerGenerator = /** @class */ (function () {
    function BarcodeMarkerGenerator(matrixTypeId, value) {
        this.typeDesc = BARCODE_MATRIX_TYPES.find(function (x) { return x.id === matrixTypeId; });
        if (!this.typeDesc)
            throw new Error("unknown barcode matrix type id: ".concat(matrixTypeId));
        if (!Number.isInteger(value))
            throw new Error("barcode value is not an integer: ".concat(value));
        if (value < 0 || value >= this.typeDesc.maxNumMarkers)
            throw new Error("barcode value out of range: ".concat(value));
        this.value = value;
        this.valueEncoded = this.typeDesc.encoder(value);
    }
    BarcodeMarkerGenerator.prototype.asSVG = function () {
        var svgStr = '<?xml version="1.0" encoding="utf-8"?>' +
            '<svg version="1.1" xmlns="http://www.w3.org/2000/svg" width="100" height="100">' +
            '<rect x="0" y="0" width="100" height="100" fill="black" />';
        var pixelSize = 50.0 / this.typeDesc.matrixSize;
        for (var y = 0; y < this.typeDesc.matrixSize; ++y)
            for (var x = 0; x < this.typeDesc.matrixSize; ++x)
                if (!this.valueEncoded[y * this.typeDesc.matrixSize + x])
                    svgStr += "<rect x=\"".concat(25 + x * pixelSize, "\" y=\"").concat(25 + y * pixelSize, "\" width=\"").concat(pixelSize, "\" height=\"").concat(pixelSize, "\" fill=\"white\" />");
        svgStr += '</svg>';
        return svgStr;
    };
    BarcodeMarkerGenerator.prototype.asSVGDataURI = function () {
        return "data:image/svg+xml,".concat(encodeURIComponent(this.asSVG())
            .replace(/'/g, '%27')
            .replace(/"/g, '%22'));
    };
    BarcodeMarkerGenerator.getMatrixTypes = function () {
        return BARCODE_MATRIX_TYPES;
    };
    return BarcodeMarkerGenerator;
}());
exports.BarcodeMarkerGenerator = BarcodeMarkerGenerator;
