// arjs-studio-backend/webpack.config.js
'use strict';

const path = require('path');
const TerserPlugin = require('terser-webpack-plugin');

module.exports = function (env) {
  const production = env.production;

  return {
    mode: production ? 'production' : 'development',

    devtool: production ? 'source-map' : 'cheap-module-eval-source-map',
    resolve: {
      // Add '.ts' and '.tsx' as resolvable extensions.
      extensions: ['.webpack.js', '.web.js', '.ts', '.tsx', '.js', '.d.ts'],
      alias: {
        '@types': path.resolve(__dirname, './src/types/index.d.ts'),
      },
    },

    entry: {
      'arjs-studio-backend': './src/index.ts',
      'arjs-studio-backend.min': './src/index.ts',
    },

    output: {
      path: path.resolve(__dirname, './dist'),
      filename: '[name].js',
      library: 'ARjsStudioBackend',
    },

    // optimization: production
    //   ? {
    //       minimize: true,
    //       minimizer: [
    //         new TerserPlugin({
    //           include: /\.min\.js$/,
    //           sourceMap: true,
    //           extractComments: false,
    //           /*terserOptions: {
    //                         module: true
    //                     }*/
    //         }),
    //       ],
    //     }
    //   : {},

    module: {
      rules: [
        {
          test: /\.(js|ts)x?$/,
          exclude: /node_modules/,
          use: {
            loader: 'babel-loader',
            options: {
              presets: [
                [
                  '@babel/preset-env',
                  {
                    targets: [
                      '> 1%',
                      'last 2 versions',
                      'not ie <= 8',
                      'not dead',
                    ],
                  },
                ],
                '@babel/preset-typescript',
              ],
              plugins: ['@babel/plugin-transform-runtime'],
            },
          },
        },
        {
          test: /\.handlebars$/,
          loader: 'handlebars-loader',
        },
        {
          test: /\.ts$/,
          loader: 'ts-loader',
          exclude: /node_modules/,
        },
      ],
    },

    stats: {
      colors: true,
      chunks: false,
      modules: false,
    },
  };
};
