// arjs-studio-backend/dist/modules/package/Package.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Package = exports.PACKAGE_GITHUB = exports.PACKAGE_ZIP = exports.AR_NTF = exports.AR_LOCATION = exports.AR_PATTERN = exports.AR_BARCODE = exports.ASSET_VIDEO = exports.ASSET_AUDIO = exports.ASSET_IMAGE = exports.ASSET_3D = void 0;
/// @ts-nocheck
var helpers_ts_1 = require("../../helpers.ts");
var index_1 = require("../../index");
exports.ASSET_3D = '3d';
exports.ASSET_IMAGE = 'image';
exports.ASSET_AUDIO = 'audio';
exports.ASSET_VIDEO = 'video';
exports.AR_BARCODE = 'barcode';
exports.AR_PATTERN = 'pattern';
exports.AR_LOCATION = 'location';
exports.AR_NTF = 'ntf';
exports.PACKAGE_ZIP = 'zip';
exports.PACKAGE_GITHUB = 'github';
/**
 * @typedef AssetParam
 * @property {Number} scale
 * @property {{width: Number, height: Number, depth: Number}} size
 * @property {Array<{latitude: Number, longitude: Number}>} locations - an array of latitude/longitude locations, for location based AR
 */
var defaultAssetParam = {
    scale: 1.0,
    size: {
        width: 1.0,
        height: 1.0,
        depth: 1.0,
    },
};
var Package = /** @class */ (function () {
    /**
     * @param {Object} config
     * @param {string} config.arType - one of barcode, pattern, location or ntf (see exported constants)
     * @param {string} config.assetType - one of 3d, image, audio or video (see exported constants)
     * @param {string|Blob} config.assetFile - the file to be show in AR
     * @param {string} config.assetName - the file name, to be included in HTML template
     * @param {AssetParam} [config.assetParam] - parameters of AR asset
     * @param {string} [config.markerPatt] - the marker image patt file (required for pattern and location AR type)
     * @param {string} [config.matrixType] - the barcode matrix type (see {@link BarcodeMarkerGenerator} exported constants, required for barcode AR type)
     * @param {number} [config.markerValue] - the barcode value of the marker (required for barcode AR type)
     */
    function Package(config) {
        this.arType = config.arType;
        this.assetType = config.assetType;
        this.assetFile = config.assetFile;
        this.assetName = config.assetName;
        this.assetParam = config.assetParam || defaultAssetParam;
        this.config = config;
    }
    /**
     * @param {Object} config - see {@link ZipProvider#serveFiles} or {@link GithubProvider#serveFiles} for available parameters
     * @param {string} config.packageType - either 'zip' or 'github'
     * @return {Promise<string|Array<number>|Uint8Array|ArrayBuffer|Blob|Buffer>} - the package, see {@link ZipProvider#serveFiles} or {@link GithubProvider#serveFiles} for actual return value
     */
    Package.prototype.serve = function (config) {
        return __awaiter(this, void 0, void 0, function () {
            var provider, generatedHtml;
            return __generator(this, function (_a) {
                provider = this.initProvider(config.packageType);
                generatedHtml = '';
                // generate HTML and add marker files, depending on chosen AR experience type
                switch (this.arType) {
                    case exports.AR_BARCODE:
                        if (!this.config.matrixType) {
                            throw new Error('Barcode-based AR needs a matrix type');
                        }
                        if (!this.config.markerValue) {
                            throw new Error('Barcode-based AR needs a marker value');
                        }
                        generatedHtml = index_1.MarkerModule.generateBarcodeHtml(this.config.matrixType, this.config.markerValue, "assets/".concat(this.assetName));
                        // barcode requires no marker file
                        break;
                    case exports.AR_PATTERN:
                        generatedHtml = index_1.MarkerModule.generatePatternHtml(this.assetType, this.assetParam, "assets/".concat(this.assetName));
                        if (!this.config.markerPatt) {
                            throw new Error('Missing marker.patt file');
                        }
                        provider.addFile('assets/marker.patt', this.config.markerPatt);
                        break;
                    case exports.AR_LOCATION:
                        generatedHtml = index_1.LocationModule.generateHtml(this.assetType, this.assetParam, "assets/".concat(this.assetName));
                        break;
                    case exports.AR_NTF:
                        throw new Error('NTF template is not implemented');
                    default:
                        throw new Error("Unknown AR type: ".concat(this.arType));
                }
                provider.addFile('index.html', generatedHtml);
                this.addAssetToProvider(provider);
                return [2 /*return*/, provider.serveFiles(config)];
            });
        });
    };
    /**
     * @param {string} packageType - either 'zip' or 'github'
     * @return {Provider}
     */
    Package.prototype.initProvider = function (packageType) {
        /** @type {Provider} */
        var provider = null;
        // init provider
        switch (packageType) {
            case exports.PACKAGE_ZIP:
                provider = new index_1.ZipProvider();
                break;
            case exports.PACKAGE_GITHUB:
                provider = new index_1.GithubProvider();
                break;
            default:
                throw new Error("Unknown provider: ".concat(packageType));
        }
        return provider;
    };
    Package.prototype.addAssetToProvider = function (provider) {
        switch (this.assetType) {
            case exports.ASSET_3D:
            case exports.ASSET_IMAGE:
                provider.addFile("assets/".concat(this.assetName), this.assetFile, index_1.ENC_BASE64);
                break;
            case exports.ASSET_AUDIO:
            case exports.ASSET_VIDEO:
                if (provider instanceof index_1.ZipProvider) {
                    provider.addFile("assets/".concat(this.assetName), this.assetFile, index_1.ENC_BINARY);
                }
                else if (provider instanceof index_1.GithubProvider) {
                    if (!(this.assetFile instanceof ArrayBuffer)) {
                        throw new Error('Audio/video asset files for GitHub must be converted to ArrayBuffer');
                    }
                    // GitHub API does not support binary files
                    // assetFile should be a base64-encoded string
                    provider.addFile("assets/".concat(this.assetName), (0, helpers_ts_1.arrayBufferToBase64)(this.assetFile), index_1.ENC_BASE64);
                }
                break;
            default:
                throw new Error("Unknown asset type: ".concat(this.assetType));
        }
    };
    return Package;
}());
exports.Package = Package;
